akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,256 @@
+
+/**
+ * Join the specified seed nodes without defining them in config.
+ * Especially useful from tests when Addresses are unknown before startup time.
+ *
+ * An actor system can only join a cluster once. Additional attempts will be ignored.
+ * When it has successfully joined it must be restarted to be able to join another
+ * cluster or to join the same cluster again.
+ */
+final case class JoinSeedNodes(seedNodes: immutable.Seq[Address]) extends ClusterCommand
+
+/**
+ * Send command to issue state transition to LEAVING for the node specified by 'address'.
+ * The member will go through the status changes [[MemberStatus]] `Leaving` (not published to
+ * subscribers) followed by [[MemberStatus]] `Exiting` and finally [[MemberStatus]] `Removed`.
+ *
+ * Note that this command can be issued to any member in the cluster, not necessarily the
+ * one that is leaving. The cluster extension, but not the actor system or JVM, of the
+ * leaving member will be shutdown after the leader has changed status of the member to
+ * Exiting. Thereafter the member will be removed from the cluster. Normally this is
+ * handled automatically, but in case of network failures during this process it might
+ * still be necessary to set the node’s status to Down in order to complete the removal.
+ */
+final case class Leave(address: Address) extends ClusterCommand
+
+/**
+ * Send command to DOWN the node specified by 'address'.
+ *
+ * When a member is considered by the failure detector to be unreachable the leader is not
+ * allowed to perform its duties, such as changing status of new joining members to 'Up'.
+ * The status of the unreachable member must be changed to 'Down', which can be done with
+ * this method.
+ */
+final case class Down(address: Address) extends ClusterCommand
+
+/**
+ * Akka Typed Cluster API entry point
+ */
+object Cluster extends ExtensionId[Cluster] {
+
+  def createExtension(system: ActorSystem[_]): Cluster =
+    new AdapterClusterImpl(system)
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] object AdapterClusterImpl {
+
+  private def subscriptionsBehavior(adaptedCluster: akka.cluster.Cluster) = Actor.deferred[ClusterStateSubscription] { ctx ⇒
+    val cluster = Cluster(ctx.system)
+    var upSubscribers: List[ActorRef[SelfUp]] = Nil
+    if (cluster.selfMember.status != MemberStatus.Up)
+      adaptedCluster.subscribe(ctx.self.toUntyped, ClusterEvent.initialStateAsEvents, classOf[MemberUp])
+
+    Actor.immutable[AnyRef] { (ctx, msg) ⇒
+      msg match {
+
+        case Subscribe(subscriber, eventClass) ⇒
+          adaptedCluster.subscribe(subscriber.toUntyped, initialStateMode = ClusterEvent.initialStateAsEvents, eventClass)
+          Actor.same
+
+        case Unsubscribe(subscriber) ⇒
+          adaptedCluster.unsubscribe(subscriber.toUntyped)
+          Actor.same
+
+        case OnSelfUp(subscriber) ⇒
+          if (cluster.selfMember.status == MemberStatus.Up) subscriber ! SelfUp(adaptedCluster.state)
```

Comment:
I think this could be racy, since `cluster.selfMember` is updated outside this actor.
E.g. this actor has already received `MemberUp` but `cluster.selfMember.status` is still Joining, then no `SelfUp` will be sent.
I think this actor must track this by itself without looking at `cluster.selfMember`

Reply:
It's not updated from outside this actor anymore after the last push, should be safe now.